### PR TITLE
[MM-8421] Fix tiff preview - webapp

### PR DESCRIPTION
**Pull request by ejachang**
#### Summary
-Added TIFF to recognized file types show that previews show up for TIFF files
-Added TIFF to supported list under [help](https://docs.mattermost.com/help/messaging/attaching-files.html#how-to-attach-files); I wasn't sure if this should be added since TIF files still don't have previews, so I only added to the English page
-I removed the TIF constant because adding it takes away the generic thumbnail

Notes:
-TIF files still don't have previews
-The jest vscode extension seems to fail, but when I run `make test` all the tests pass. I think someone also mentioned it [here](https://community.mattermost.com/core/pl/8mnqqsstyp8xxywbwjc37kn4by)
-I wasn't sure if two items on the checklist were needed so I added the '?' next to it

-regarding the tests - would TIFF fall under "other file" in view_image.test? Would TIFF need a more specific test in image_preview.test?

[Redux PR](https://github.com/mattermost/mattermost-redux/pull/764)
[Server PR](https://github.com/mattermost/mattermost-server/pull/10170)

#### Ticket Link
[[PLT-8516](https://github.com/mattermost/mattermost-server/issues/8092)] 

#### Checklist
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [ ] Added or updated unit tests (required for all new features)
- [?] Needs to be implemented in mobile (link to PR or User Story)

### CODE REVIEW

**Comment by esethna** [comment]
@jwilander @cpanato same issue is happening on this server:

image

Must be an issue with the spinmints if you guys can help resolve

**Comment by jwilander** [comment]
@esethna fixed it on this instance, it was file system permission issue

When Carlos gets back I'll have him look at fixing the root cause

**Comment by esethna** [comment]
Ah thanks, Lev. Will wait for the server-side to get merged first.

**Comment by hanzei** [comment]
Server changes are merged.

**Comment by levb** [comment]
Right, so 2 things still needed before we test this, I think. (a) merge the redux PR (I just did that), (b) update `package.json` here to point at it, and (c) merge latest master changes.

@ejachang would you like (b) and (c) so we can QA on spinmint?

**Comment by ejachang** [comment]
@levb Yes! Will do ASAP.

**Comment by ejachang** [comment]
@levb Done with (b) and (c)!

**Comment by DHaussermann** [comment]
Hi @ejachang, I've tested this on the SpinMint and attaching the images file a post is working fine.

But, when using a link to a .tiff image such as `http://skuld.bmsc.washington.edu/raster3d/examples/ortex.tiff`, It does not generate a preview when posting the link either by itself or inside of markdown. Also, the image will  not show when using it as an attachment in post made by webhook.

tiff preview

Can you confirm if these cases should also be in-scope for this PR?

**Comment by ejachang** [comment]
@DHaussermann @esethna Yes! will do.

**Comment by hmhealey** [comment]
@DHaussermann These changes will only affect TIFFs uploaded as file attachments, so it's expected that they won't render in image previews, markdown images, or message attachments. The big empty boxes are definitely something we should look at fixing though

**Comment by levb** [comment]
@DHaussermann taking @hmhealey's comment into account, so can we consider this verified by QA

**Comment by DHaussermann** [comment]
@levb  and @hmhealey I understand that there is no need to support the link previews in the scope of this PR. But, the empty space created when the preview does not render (see screenshot) is new behavior and a regression from master. 

Can we make change so that the app does not attempt to render image previews for this file extension? If need be, I can create I separate task to address this.

**Comment by esethna** [comment]
Ah I see, thanks for the catch Dylan. We need to either:

1. Support the preview in image previews, markdown images, or message attachments
2. Not rendering the large blank spaces

This is definitely not shippable as is given it's exposing broken/partially built functionality. @hmhealey @ejachang what would you propose as the next step here?

image

**Comment by ejachang** [comment]
@DHaussermann @esethna I think the blank spaces are happening for the link preview tested because there are problems decoding the TIFF/TIF file. On the server side when this happens the decoding of the TIFF/TIF is  `nil`. Using the golang TIFF package doesn't solve the problem as it gives the same error. 
I noticed there was this issue with TIFs as opposed to all the TIFFs I tested on, so I thought there were no issues decoding TIFFs. That is why I left the generic preview for TIFs, but I guess this decoding issue also applies to TIFFs as well. When TIFF/TIFs are decoded no blank space/preview pops up when posting links.
screen shot 2019-03-04 at 11 31 57 pm
I'm a bit new to handling/decoding TIFF/TIFs so I'm still trying to figure out the reason why there's this issue of decoding certain TIFF/TIFs, e.g. different file organization in different TIFF/TIFs. I was actually wondering until this is fixed, if in the cases where the app cannot decode the TIFF/TIF whether it is ok for the app to render a generic preview thumbnail with a message that the TIFF/TIF can't be decoded at this time?

**Comment by hmhealey** [comment]
To get rid of the blank space, we'll have to separate which file extensions are used to show previews for file attachments and which are used for image link previews. Currently, they both use the IMAGE_TYPES constant in constants.jsx, so we'll need to split those out.

I don't think this is related to the server side changes or any decoding issues there since only affect file attachments and not image link previews.

**Comment by levb** [comment]
@hmhealey how do you suggest we move forward? I am confused - is this "blank space" issue specific to TIFF files, and if such should probably fixed in or together with this PR? Also, if it is TIFF extension or mime-type specific, can you please elaborate more on what makes TIF special?

If it is not TIFF-specific, then seems like we should open a new ticket for it, and fix it separately?

**Comment by ejachang** [comment]
Yes, I forgot links to TIFs are not affected because they are not included in the constants in this PR. Sorry for the confusion from my last comment. I thought it was related to the server because the thumbnail preview of the Ortex TIFF file wasn't being decoded and I had been focused on that for a while. 

If there is an issue with decoding certain TIFF files though, like with the Ortex TIFF file, would this also need to be addressed? At first I thought there were no issues decoding TIFF files because I had trouble testing the TIF files I had and no problem with any TIFF files. I see now instead certain TIF and TIFF files are able to be decoded, but some aren't.

**Comment by hmhealey** [comment]
I'd prefer to show the generic thumbnail if it fails to decode, but we don't actually do this when the server fails to generate thumbnails for other files, so I think we should create a separate ticket to address that for all images, including TIFFs.

For example, if I rename a text file to have a `.png` extension, the app gets into a weird state
image

**Comment by hmhealey** [comment]
This PR has kind of stalled since I think we're all confused about where it's at right now. If I'm understanding correctly, there's two similar issues blocking this PR:
1. When someone posts `http://skuld.bmsc.washington.edu/raster3d/examples/ortex.tiff`, it causes an empty rectangle to appear. This is a new issue caused by this PR. It's happening because [the PostBodyAdditionalContent component](https://github.com/mattermost/mattermost-webapp/blob/master/components/post_view/post_body_additional_content/post_body_additional_content.jsx#L166) uses the same [`IMAGE_TYPES`](https://github.com/mattermost/mattermost-webapp/pull/2333/files#diff-4dd3862e12ff806bd050d21238a9139cR604) constant that is used to determine whether or not to show a file attachment thumbnail. To fix this, we should have the PostBodyAdditionalContent use a separate array that doesn't include `.tiff` since most browsers can't render them.

    This issue is relatively easy to fix, and it should be fixed before merging this since it's easy to hit by mistake (eg just paste a link ending in `.tiff`).
2. When a message attachment or markdown image is posted linking to a `.tiff` file, it causes an empty rectangle to appear. This is happening because of conflicting logic between [the server PR to add tiff support](https://github.com/mattermost/mattermost-server/pull/10170) and some assumptions we made with the image dimensions part of the post metadata changes.

    As part of the metadata changes, we intentionally made it so that images that failed to load would keep their original dimensions to avoid causing the page content to jump around. Because of the server PR to add tiff support, the server now sends dimensions for tiff images to the client, and then the client renders the empty rectangle because it can't show a tiff.

    The fix for this isn't as simple as the first one because message attachments and markdown images don't follow that `IMAGE_TYPES` array at all, so they'd require additional logic to determine that the link doesn't actually point to an image. Since this is less likely to happen and because it's a bigger change, I think we should file a separate ticket for this and address it separately.

Thoughts? @ejachang @levb @esethna

**Comment by ejachang** [comment]
I agree that it makes sense to do #1 for the scope of this PR. Also because in regards to #2, it sounds like if any other image fails to load an empty rectangle will appear for those images, and not just TIFFs?

**Comment by hmhealey** [comment]
Yeah, that's correct about #2. It's intended to stop the page layout from changing and jumping around after loading the post at the expense of a bit of weirdness when the image fails to load

**Comment by levb** [comment]
Complete agreement re: issue1. Thanks @hmhealey for taking the time and also explaining it to me offline.

@hmhealey @ejachang maybe I am misunderstanding something about issue (2) but shouldn't the client be using the preview image, which should be available and is always a JPEG? The HasPreview bit in FileInfo should be set.

**Comment by hmhealey** [comment]
The naming is confusing, but issue 2 affects message attachments (aka Slack attachments), not file attachments.

**Comment by ejachang** [comment]
Is it okay to move forward for this PR with fixing issue 1 and separating the other issues in separate tickets since it seems like those issues encompass all images, and not only tiffs? i.e. (A) for issue 2/message attachments and (B) when the server fails to generate images for files.

**Comment by hmhealey** [comment]
@ejachang Yeah, that's correct. If you can fix issue 1, I'll go file a ticket for issue 2 that can be fixed  separately.

File: package-lock.json [hunk 1]
```diff
@@ -10784,7 +10784,7 @@
     },
     "ora": {
       "version": "0.2.3",
-      "resolved": "http://registry.npmjs.org/ora/-/ora-0.2.3.tgz",
+      "resolved": "https://registry.npmjs.org/ora/-/ora-0.2.3.tgz",
```
Comment (ejachang) [inline package-lock.json]: I accidentally pushed some changes to the package-lock.json that I didn't mean to. This is to fix it. Sorry for any confusion!

**Comment by amyblais** [comment]
Looks like this doesn't make it for today's Feature Complete. Is the redux PR going to cause any issues as it's already merged to master without this PR?

**Comment by ejachang** [comment]
Hi @amyblais I don’t think the redux PR is going to cause issues because the change in that PR was for the IMAGE_TYPES constant, but it seems the web app still uses the IMAGE_TYPES constant from mattermost-webapp/utils/constants.jsx and not from redux.
I realize now the addition in the redux PR should have been done when migrating, unless I am misunderstanding something? If that is correct, would it be okay for the change to remain for future migration? 
Sorry about this and the late reply!

**Comment by ejachang** [comment]
@amyblais to edit/update on my previous comment, I think the redux PR affects the mobile app. When I post a link on the public test channel it only shows the link in the webapp, but on the mobile app (iphone 7s)  an empty rectangle appears, which is currently what is going on with[ issue (1)](https://github.com/mattermost/mattermost-webapp/pull/2333#issuecomment-471773551). Should the redux PR be reverted?

**Comment by amyblais** [comment]
@jwilander @hmhealey Do you know if this should be reverted from v5.10? The redux and server PRs were merged to v5.10 but webapp PR is not. Based on this comment https://github.com/mattermost/mattermost-webapp/pull/2333#issuecomment-480568841 this seems to affect the mobile app.

**Comment by jwilander** [comment]
Yes, that might affect the mobile app. I'll revert for the release-5.10 branch

**Comment by esethna** [comment]
@hmhealey @ejachang checking in, what are next steps on this PR?

**Comment by hmhealey** [comment]
@esethna I was waiting on PM/QA review before doing my review since the Dev Review label wasn't assigned yet. The code changes are small, so I can go do my dev review right now though

File: components/post_view/post_body_additional_content/post_body_additional_content.test.jsx [hunk 2]
```diff
@@ -50,6 +50,27 @@ describe('components/post_view/PostBodyAdditionalContent', () => {
         });
     });
 
+    test('isLinkImage, should return false for tiffs', () => {
+        const testCases = [
+            {link: 'http://localhost/some_image1234.tiff', result: false},
```
Comment (hmhealey) [inline components/post_view/post_body_additional_content/post_body_additional_content.test.jsx]: These could probably go with the other tests for `isLinkImage`, but that doesn't really matter much.

Comment (ejachang) [reply components/post_view/post_body_additional_content/post_body_additional_content.test.jsx]: @hmhealey you're right; I put these tests with the others in `isLinkImage`!

**Comment by hanzei** [comment]
@esethna Would you please review when you have the time?

**Comment by ejachang** [comment]
@esethna looking into it - after the changes in March it seemed to work okay. Not sure what happened. 
Screen Shot 2019-05-01 at 1 49 41 AM
Currently it seems like the box shows for admin users, but seems okay for regular users. 
Screen Shot 2019-05-01 at 12 11 26 AM

**Comment by ejachang** [comment]
Looks like the preview is showing for admin users because both are true for admins:
I) [`isEmbedVisible`](https://github.com/mattermost/mattermost-webapp/blob/3d29771a55c825ebfb7abeb04bc34686da983fce/components/post_view/post_body_additional_content/post_body_additional_content.jsx#L335) - On the regular user side embedded images are being passed as false, so there are no previews that show in posts for tiffs.
II) isMetadataImage - this was to allow image previews for shortened links. But because [it allows image file types to generate preview images,](https://github.com/mattermost/mattermost-webapp/blob/3d29771a55c825ebfb7abeb04bc34686da983fce/components/post_view/post_body_additional_content/post_body_additional_content.jsx#L76)  it is generating a preview for [toggeable embed images](https://github.com/mattermost/mattermost-webapp/blob/3d29771a55c825ebfb7abeb04bc34686da983fce/components/post_view/post_body_additional_content/post_body_additional_content.jsx#L238). This includes tiffs even though it can't show it. I’m not sure if the best way to move forward is to flag isMetadataImage as false if the link points to a tiff image?

**Comment by hanzei** [comment]
@hmhealey Would you please take a look a the above questions? https://github.com/mattermost/mattermost-webapp/pull/2333#issuecomment-489291576

**Comment by hanzei** [comment]
@hmhealey Would you please take a look a the above questions? https://github.com/mattermost/mattermost-webapp/pull/2333#issuecomment-489291576

**Comment by hmhealey** [comment]
Since we added tiff support to the server, `image.DecodeConfig` on the server now works for tiff images anywhere that it's called, and [we use that to identify image links](https://github.com/mattermost/mattermost-server/blob/master/app/post_metadata.go#L499). I think it probably makes the most sense to handle this on the server instead of the client since it's what changed here.

@ejachang Would you be able to submit a PR to add a special case to [`parseImages` in `app/post_metadata.go`](https://github.com/mattermost/mattermost-server/blob/master/app/post_metadata.go#L493) on the server so that it just returns `nil` when the format returned from `DecodeConfig` is tiff? That'll fix the issue with `isMetadataImage` since `embed.type` won't be `image` any more.

**Comment by ejachang** [comment]
@hmhealy ah I see! Yes, that makes more sense. Will do!

**Comment by ejachang** [comment]
@hmhealey  please excuse the late response. I was starting to work on the special case PR, but I noticed a few recent changes. 
- Empty boxes were not showing up for tiff links anymore. I wasn't sure if it was because of [this PR](https://github.com/mattermost/mattermost-webapp/pull/2638), since at this time `isMetadataImage` was false for tiff links and `post.metadata` was empty. 
- After updating today though, all image previews for links were not showing up.
Screen Shot 2019-05-13 at 23 23 20
I wanted to check that even with these changes if it would be fine to still move forward with the special case PR?

**Comment by hmhealey** [comment]
@ejachang Can you update to the latest master and see how it works? That PR caused some bugs with the visibility of attachments, but those should be fixed now

**Comment by esethna** [comment]
@hmhealey @ejachang should we add the work in progress label to this PR?

**Comment by ejachang** [comment]
@hmhealey I'm still don't see the image previews for the image links. If I understand what happened correctly it is because: 
1) `post.metadata` is empty so `getEmbed` returns `null` and doesn't render the image preview
2) `Utils.extractFirstLink(props.post.message)` was [taken out](https://github.com/mattermost/mattermost-webapp/commit/c563d187a00e2a4893eabc8c52334a5e839c6fe8) of utils and the component.
After this [PR](https://github.com/mattermost/mattermost-webapp/pull/2638) image links were still visible even though `props.metadata` was empty I think because the link was extracted from `Utils.extractFirstLink(props.post.message)`.

I'm still figuring what is causing `post.metadata` to be empty since I'm not too familiar with that part. Would you be able to advise? 

@esethna I think the WIP label would be fine to add to this PR.

**Comment by hmhealey** [comment]
Oh, I missed that `post.metadata` was empty. That should be populated as long as the server is new enough (5.8+ supported it, although it was off by default until 5.10) and as long as `DisablePostMetadata` is false in the config. Also, `EnableLinkPreviews` has to be true for `post.metadata.embeds` to be non-empty.

If the server is new enough and those settings are set correctly, do you see any errors in your server logs? The server has to be able to request those web pages/images itself, so it could be a firewall issue

**Comment by ejachang** [comment]
@hmhealey setting `EnableLinkPreviews` to true in either the server config fixed it for the sysadmin. Is there a reason why it's initially set to false? 

I didn't see any errors in my server logs related to the this PR other than setting SiteURL. I've also created the special server PR so image previews shouldn't show for tiff image links.

**Comment by hmhealey** [comment]
Oh right, I forgot about that setting. I think it defaulted to false back when link previews were still in beta, so it probably was just left that way.

**Comment by esethna** [comment]
Received the following error after trying to drag and drop a TIFF image: 
image

**Comment by hmhealey** [comment]
A 502 error sounds like an issue with the Spinmint. Can you send me the tiff file you're using and I can test it locally?

I am having issues getting certain TIFF files to work with Go's TIFF parsing library. `http://skuld.bmsc.washington.edu/raster3d/examples/ortex.tiff` that Dylan was using above fails to parse with a "not enough pixel data" error. The first few ones on [this page](http://www.fileformat.info/format/tiff/sample/) don't work because of an "unsupported feature: compression value 4" error. When this happens, you get a blank preview image. That happens on master as well when we fail to generate an image preview, but that only currently happens in extreme circumstances.

@esethna Do you think this is something we should fix now, or is it acceptable to file a separate ticket for that? I don't know how often this will happen in practice, but this is how it looks
image

As long as the server can properly decode the gif, this looks good though
image

**Comment by esethna** [comment]
@hmhealey if it happens in extreme cases I think it's not necessary to block this PR on it

**Comment by hmhealey** [comment]
@esethna That tiff file works fine for me locally, so I think it's safe to blame your issue on the spinmint.

As for the other tiff files, I don't know how commonly it happens. I don't have much experience with working with those. Perhaps we just merge this and then have QA test it with a variety of tiff files? It doesn't look that bad if we're unable to generate thumbnails, and a similar issue already happens if you upload a non-image file with an image extension anyway.

**Comment by esethna** [comment]
Added to the QA PR review queue

**Comment by jwilander** [comment]
The test server above is still working

**Comment by hanzei** [comment]
@hmhealey @jwilander What are the next steps on this PR?

**Comment by ejachang** [comment]
Is there anything I should do for this PR? Seems like things are okay and the QA review label was up so I wasn't sure.

**Comment by hmhealey** [comment]
Sorry, looks like this PR got lost on my review queue. I'll add someone from QA for review since they seem to have lost it as well.

Last time I reviewed, I know there were some issues with that one TIFF file that we were using to test, but it seems like an isolated issue as best as I can tell, so it's nothing that would block this PR. Any fixes for that would affect the server anyway, not the web app.

**Pull request merged**